[TACHYON-786] Login Mechanism: Add CUSTOM_LOGIN (by JAAS framework)

ooq opened the pull request:
JIRA: [TACHYON-786](https://tachyon.atlassian.net/browse/TACHYON-786)

In production use of Tachyon, we want different applications to have separate user identities, even if they are running under same operating system user. We should allow this by letting the application to specify their username, through Java configuration parameter. This JIRA enables that in the SIMPLE login mechanism. If tachyon.security.username is specified by the application and is not empty, the login modules uses tachyon.security.username as the username; otherwise, we fall back to OS_SPECIFIC_LOGIN.

ooq commented:
@dongche @shenguoquan @apc999 Could you review this PR? Thanks.

apc999 commented on common/src/main/java/tachyon/security/CustomLoginModule.java:
```diff
@@ -0,0 +1,118 @@
+   */
+  @Override
+  public boolean login() throws LoginException {
+    String userName = System.getProperty(Constants.TACHYON_SECURITY_USERNAME, "");
```
Can you use `TachyonConf` to get `Constants.TACHYON_SECURITY_USERNAME` rather than `System.getProperty`

apc999 commented on common/src/test/java/tachyon/security/LoginUserTest.java:
```diff
@@ -57,6 +57,39 @@ public void getSimpleLoginUserTest() throws Exception {
+  public void getCustomLoginUserTest() throws Exception {
+    TachyonConf conf = new TachyonConf();
+    conf.set(Constants.TACHYON_SECURITY_AUTHENTICATION, "SIMPLE");
+    System.setProperty(Constants.TACHYON_SECURITY_USERNAME, "tachyon-user");
```
Also through `TachyonConf`?

ooq commented:
Discussed with @apc999 , we will use `System.getProperty()` for now and move to `TachyonConf` after `TachyonConf` is refactored to Singleton.

dongche commented:
@ooq LGTM. Thanks for figuring out this problem, and the solution is good.

Only one minor comments is also marking the System.setProperty in the test as TODO, in case missing updating this test part in the future work. Since we already have TODO in the code, this is not a big deal. Up to you.

apc999 commented on common/src/main/java/tachyon/security/CustomLoginModule.java:
```diff
@@ -0,0 +1,120 @@
+import tachyon.Constants;
+
+/**
+ * A custom login module that creates a user based on the user name provided through application
```
"based on the user name provided through application configuration": can you be more specific, e.g., specifying which property, maybe through an example.

apc999 replied:
And can you specify a little bit here in javadoc why this module can be useful (e.g., in what context)?

apc999 commented on common/src/main/java/tachyon/security/CustomLoginModule.java:
```diff
@@ -0,0 +1,120 @@
+    //TODO: after TachyonConf is refactored into Singleton, we will use TachyonConf
+    //instead of System.getProperty for retrieving user name.
+    String userName = System.getProperty(Constants.TACHYON_SECURITY_USERNAME, "");
+    if (userName != "") {
```
can you use `!userName.empty()`

apc999 commented on common/src/main/java/tachyon/security/CustomLoginModule.java:
```diff
@@ -0,0 +1,120 @@
+  /**
+   * Retrieve the user name by querying the property of Constants.TACHYON_SECURITY_USERNAME.
+   *
+   * @return true if customized user name is not null or empty
```
"@return true if customized user name is not null or empty", this statement is misleading, as I thought it returns true if ... is empty, which is not the case.

apc999 commented on common/src/test/java/tachyon/security/LoginUserTest.java:
```diff
@@ -57,6 +57,39 @@ public void getSimpleLoginUserTest() throws Exception {
+    User loginUser = LoginUser.get(conf);
+
+    Assert.assertNotNull(loginUser);
+    Assert.assertFalse(loginUser.getName().isEmpty());
```
Is there any way to assert that the userName is really OS user? I guess I want to make sure that, in this setting, the login module does fallback to the second choice.

dongche replied:
Seems the only way is we use the login module in this test again to get OS user as the expected value.

ooq replied:
Would not `System.getProperty("user.name")` work? I am now retrieving OS user using this property, let me know if it would not in certain scenarios.

apc999 commented:
@ooq , thanks for the patch. I left a few minor comments. It looks in general good.

ooq commented:
@apc999 @dongche Thanks for the feedback and apologize for the delay. 
I have addressed all the issues. For @apc999 's comment on checking the login user against the OS user, I am now using `System.getProperty("user.name")` to retrieve the latter, let me know if that would not work.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/Tachyon-Pull-Request-Builder/4421/
Test PASSed.

shenguoquan commented:
LGTM +1. Thanks @ooq

dongche commented on common/src/test/java/tachyon/security/LoginUserTest.java:
```diff
@@ -57,6 +57,43 @@ public void getSimpleLoginUserTest() throws Exception {
     Assert.assertFalse(loginUser.getName().isEmpty());
```
Could you please also help to update this check to compare the user name, like below test? Thanks. 
It looks better to check the equality instead of just not empty. I did not realize this at that time.

dongche commented:
@ooq Nice idea. I think it is ok to use that system property. The `user.name` property should have no chance to be changed in test execution.

Original `getSimpleLoginUserTest` could be also checked like this. Could you help to update it in this PR? Thanks.

ooq commented:
@dongche Good thought on updating the simple login tests. But given these are code logic changes that are not related to this PR, we should probably update it in separate PR.

dongche commented:
@ooq Sure, I will open another PR for this.

LGTM. Thanks.

apc999 commented:
+1 LGTM
